### PR TITLE
test(Forms): Add vanilla example coverage

### LINKED ISSUE #2893: Expand on form Storybook e2e examples: vanilla

### PR DESCRIPTION
**Pull request by m7kvqbe1**
## Related issue

Closes #2893

## Overview

Consume form components without relying on any companion libraries (Formik, react-hook-form etc).

## Reason

>We need to be sure that the form components work in a variety of use cases.

## Work carried out

- [x] Create new vanilla forms example
- [x] Tweak some types and code style

### CODE REVIEW

File: packages/react-component-library/src/forms/vanilla/useVanillaForm.ts [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+  initialErrors: FormErrors,
+  validation: Record<string, (value: string) => boolean>
+) => {
+  const [formErrors, setFormErrors] = useState<FormErrors>(initialErrors)
+  const [formPayload, setFormPayload] = useState<T | undefined>()
+  const [formState, setFormState] = useState<T>(initialState)
+  const [isSubmitting, setIsSubmitting] = useState<boolean>(false)
+
+  const handleChange = (e: React.FormEvent | SyntheticFormEvent): void => {
+    const { name, value } = e.target as HTMLInputElement
```
Comment (jpveooys) [inline packages/react-component-library/src/forms/vanilla/useVanillaForm.ts]: Are we able to use `React.FormEvent<HTMLInputElement>` and `e.currentTarget` here if `SyntheticFormEvent` is changed to match?

File: packages/react-component-library/src/forms/vanilla/useVanillaForm.ts [hunk 2]
```diff
@@ -0,0 +1,99 @@
+import React, { useState } from 'react'
```
Comment (jpveooys) [inline packages/react-component-library/src/forms/vanilla/useVanillaForm.ts]: Would be nice to exclude this file from SonarQube coverage if we're able to.

File: packages/react-component-library/src/forms/vanilla/vanilla.stories.tsx [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+
+      <pre data-testid="form-example-values">
+        {JSON.stringify(formPayload, null, 2)}
+      </pre>
+    </main>
+  )
+}
+
+export default {
+  title: 'Forms/Vanilla',
```
Comment (jpveooys) [inline packages/react-component-library/src/forms/vanilla/vanilla.stories.tsx]: I wonder if 'Native' would be more descriptive than 'Vanilla'...

File: packages/react-component-library/src/forms/vanilla/vanilla.stories.tsx [hunk 4]
```diff
@@ -0,0 +1,194 @@
...
+  component: ExampleVanillaForm,
+} as Meta
```
Comment (jpveooys) [inline packages/react-component-library/src/forms/vanilla/vanilla.stories.tsx]: Btw they've changed the types a bit in Storybook 6.4:

image

But we need to update it everywhere...

**Review by jpveooys — COMMENTED** [review]
Nice!

Few other points...

1. Noticed something a bit weird – in some cases the JSON data updates without the form being submitted:

    vanilla

    Any ideas?

2. There seems to be two identical entries in the Storybook sidebar:

    image

    Any way to get rid of one?
    
3. Would be good to keep the initial error state of the email field consistent with the other examples if possible.

File: packages/react-component-library/src/forms/vanilla/vanilla.stories.tsx (hunk 4 above)
Comment (m7kvqbe1) [reply packages/react-component-library/src/forms/vanilla/vanilla.stories.tsx]: Updated for forms and created a [new ticket](https://github.com/defencedigital/mod-uk-design-system/issues/2961) to handle this across the rest of the codebase.

File: sonar-project.properties [hunk 5]
```diff
@@ -7,8 +7,7 @@ sonar.projectName=mod-uk-design-system
 # Path to sources
 sonar.sources=packages/react-component-library/src
 sonar.exclusions=**/*.test.ts,**/*.test.tsx*,**/*.stories.tsx*
-sonar.coverage.exclusions=**/*.test.ts,**/*.test.tsx*,**/*.stories.tsx*
-
+sonar.coverage.exclusions=**/*.test.ts,**/*.test.tsx*,**/*.stories.tsx*,forms/**/*
```
Comment (jpveooys) [inline sonar-project.properties]: Might need to be the full path:

```suggestion
sonar.coverage.exclusions=**/*.test.ts,**/*.test.tsx*,**/*.stories.tsx*,packages/react-component-library/src/forms/**/*
```

File: packages/react-component-library/cypress/specs/forms/index.spec.ts [hunk 6]
```diff
@@ -42,6 +52,10 @@ describe('Form Examples', () => {
           name: 'react-hook-form',
           uri: '/iframe.html?id=forms-react-hook-form--default&viewMode=story',
```
Comment (jpveooys) [inline packages/react-component-library/cypress/specs/forms/index.spec.ts]: These URLs need updating now I think...

**Review by jpveooys — APPROVED** [review]
:+1:

**Pull request merged**